Acme tls mode nixos 24.05

nobodyinperson opened the pull request:
# See [the real PR in nixpkgs](https://github.com/NixOS/nixpkgs/pull/340136), this is just for me internally

## Description of changes

This PR adds `services.acme.[...].tlsMode` and `tlsPort` so the TLS-ALPN challenge can be used instead of HTTP or DNS, which is necessary if port 80 is blocked and access to DNS records is impossible.

### TODO

- [ ] target `master`, not `24.05` (currently using that for testing)
- [x] make `services.acme.defaults.tls{Mode,Port}` work
- [x] conditionally allow unprivileged `acme` user to open a `tlsPort < 1024`
- [x] allow specifying systemd service names to temporarily stop while acme is opening `tlsPort` (e.g. `"nginx"`).

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Atemu commented on nixos/modules/security/acme/default.nix:
```diff
@@ -624,6 +626,18 @@ let
         '';
       };
 
+      tlsMode = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Use TLS challenge instead of HTTP.";
+      };
```
mkEnableOption

Atemu commented on nixos/modules/security/acme/default.nix:
```diff
@@ -624,6 +626,18 @@ let
+      };
+
+      tlsPort = mkOption {
+        type = types.int;
```
```suggestion
        type = types.port;
```

Atemu commented on nixos/modules/security/acme/default.nix:
```diff
@@ -222,6 +222,8 @@ let
       ++ optionals (data.dnsResolver != null) [ "--dns.resolvers" data.dnsResolver ]
     ) else if data.s3Bucket != null then [ "--http" "--http.s3-bucket" data.s3Bucket ]
     else if data.listenHTTP != null then [ "--http" "--http.port" data.listenHTTP ]
+    # TODO: why is defaults.tlsMode not applied to data.tlsMode so we have to do this ||?
```
I'd like to see this resolved. I don't know the module well enough to know what's going on here myself though.

nobodyinperson replied:
Figured it out. My options were missing that `defaultAndText` thing.

nobodyinperson commented on nixos/modules/security/acme/default.nix:
```diff
@@ -180,607 +218,719 @@ let
+
+        tlsMode = lib.mkOption {
+          type = lib.types.bool;
+          inherit (defaultAndText "tlsMode" false) default defaultText;
```
@Atemu Turns out `mkEnableOption` couldn't have been used here anyway, because of that defaults-system.

Atemu replied:
It's possible to `mkEnableOption ... // defaultsAndText ...`.

nobodyinperson replied:
Okay but there are many more occurrences of the same pattern and I don't think it makes sense to clobber this PR (this is the wrong one btw for me internally) with such refactoring.

nobodyinperson committed:
acme: add TLS ALPN challenge

This commit adds the `tlsMode`, `tlsPort` and `conflictingServices`
options to `services.acme.[...]` to allow using it even if port 80 is
blocked and DNS access is impossible.
